src/main/java/io/vertx/core/impl/VertxImpl.java

```diff
@@ -525,40 +536,44 @@ public synchronized void close(Handler<AsyncResult<Void>> completionHandler) {
 
     closeHooks.run(ar -> {
       deploymentManager.undeployAll(ar1 -> {
-        if (haManager() != null) {
-          haManager().stop();
-        }
-        addressResolver.close(ar2 -> {
-          eventBus.close(ar3 -> {
-            closeClusterManager(ar4 -> {
-              // Copy set to prevent ConcurrentModificationException
-              Set<HttpServer> httpServers = new HashSet<>(sharedHttpServers.values());
-              Set<NetServerImpl> netServers = new HashSet<>(sharedNetServers.values());
-              sharedHttpServers.clear();
-              sharedNetServers.clear();
-
-              int serverCount = httpServers.size() + netServers.size();
-
-              AtomicInteger serverCloseCount = new AtomicInteger();
-
-              Handler<AsyncResult<Void>> serverCloseHandler = res -> {
-                if (res.failed()) {
-                  log.error("Failure in shutting down server", res.cause());
+        this.executeBlocking(fut -> {
+          if (haManager() != null) {
```

Comment:
move the test outside the task so we don't execute the task when there is no need to be